packages/mgt-chat/src/statefulClient/StatefulGraphChatListClient.ts

```diff
@@ -194,13 +205,89 @@ class StatefulGraphChatListClient implements StatefulClient<GraphChatListClient>
     this._notificationClient.tearDown();
   }
 
+  private sleep(ms: number) {
+    return new Promise(resolve => setTimeout(resolve, ms));
+  }
+
+  /**
+   * Switches to a fatal error state and logs the error.
+   */
+  public raiseFatalError(e: Error) {
+    error(e);
+    this.notifyStateChange((draft: GraphChatListClient) => {
+      draft.status = 'fatal error';
+      draft.chatThreads = [];
+    });
+  }
+
   /**
    * Load more chat threads if applicable.
    */
-  public async loadMoreChatThreads() {
-    const state = this.getState();
-    const items: GraphChatThread[] = [];
-    await this.loadAndAppendChatThreads('', items, state.chatThreads.length + this.chatThreadsPerPage);
+  public async tryLoadMoreChatThreads() {
+    try {
+      // do not load if another activity is in progress
+      if (this._loadPromise || this._loadMorePromise) {
+        return;
+      }
+
+      // make sure there is something to load
+      const state = this.getState();
+      if (!state.moreChatThreadsToLoad) {
+        return;
+      }
+
+      // set promise; load and append
+      this._loadMorePromise = this.loadAndAppendChatThreads('', [], state.chatThreads.length + this.chatThreadsPerPage);
+      await this._loadMorePromise;
+    } finally {
+      this._loadMorePromise = undefined;
+    }
+  }
+
+  /**
+   * Loads chat threads without regard for the existing threads.
+   */
+  public async tryLoadChatThreads() {
+    // set the state to loading and clear all chats
+
+    // do not load if there is another load already running
+    if (this._loadPromise) {
+      return;
+    }
+
+    // wait for any load more to finish
+    if (this._loadMorePromise) {
+      await this._loadMorePromise;
+    }
+
+    // clear and announce loading
+    this.notifyStateChange((draft: GraphChatListClient) => {
+      draft.status = 'loading chats';
+      draft.chatThreads = [];
+    });
+
+    // try several times to load more chats
+    try {
+      let loaded = false;
+      let count = 0;
+      do {
+        count++;
+        try {
+          this._loadPromise = this.loadAndAppendChatThreads('', [], this.chatThreadsPerPage);
+          await this._loadPromise;
+          loaded = true;
+        } catch (e) {
+          if (count > 3) {
+            error('Failed to load chat threads; aborting...', e);
+            throw Error('Failed to load chat threads even after 3 attempts; no more attempts will be made.');
```

Comment:
why throw? we already logged this failure.

Reply:
I want it to have an exception if it fails. In ChatList it handles .catch().

Reply:
I guess what's the point? Now you are logging the error twice.